### PR TITLE
[Truffle] Start to support the Rubinius FFI and POSIX calls

### PR DESCRIPTION
**Pull request by chrisseaton**
This is the start of the work that will enable us to use the Rubinius implementation of File, IO and Socket, which are large areas of missing functionality.

@thomaswue @nirvdrum @eregon

### CODE REVIEW

File: truffle/src/main/ruby/core/rubinius/common/file.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  end
+
+  ##
+  # Deletes the named files, returning the number of names
+  # passed as arguments. Raises an exception on any error.
+  #
+  # See also Dir::rmdir.
+  def self.unlink(*paths)
+    paths.each do |path|
+      n = POSIX.unlink Rubinius::Type.coerce_to_path(path)
```
Comment (chrisseaton) [inline truffle/src/main/ruby/core/rubinius/common/file.rb]: We can call simple POSIX functions fairly directly.

File: truffle/src/main/ruby/core/rubinius/common/process.rb [hunk 2]
```diff
@@ -28,8 +28,28 @@
...
+      @maxgroups = m
+    end
+  end
+
+  def self.groups
+    g = []
+    FFI::MemoryPointer.new(:int, @maxgroups) { |p|
+      num_groups = FFI::Platform::POSIX.getgroups(@maxgroups, p)
+      Errno.handle if num_groups == -1
+      g = p.read_array_of_int(num_groups)
```
Comment (chrisseaton) [inline truffle/src/main/ruby/core/rubinius/common/process.rb]: More complex POSIX functions that do things like write into a pointer you pass them are also supported, and then end up allocating native heap using `Unsafe`.

**Comment by nirvdrum** [comment]
I haven't yet tried to run any of this, but my immediate concern is breaking Windows.  While there's an awful lot we don't yet fully support, JRuby+Truffle is substantially usable on Windows today.  I think jnr-posix will get us most of the way there and it may turn out we need to fill in some Win32 gaps there, but I wanted to raise the topic now before we get too far.  I'll try running the branch in Windows tomorrow to see where we're at.

File: truffle/src/main/ruby/core/rubinius/common/process.rb (hunk 2 above)
Comment (nirvdrum) [inline truffle/src/main/ruby/core/rubinius/common/process.rb]: Do we only care about OpenJDK if not running in Graal?  JRuby generally has fallbacks if `Unsafe` isn't available.

File: truffle/src/main/ruby/core/rubinius/platform/ffi.rb [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+  TypeSizes = Rubinius::LookupTable.new
+  TypeSizes[1] = :char
+  TypeSizes[2] = :short
+  TypeSizes[4] = :int
+  TypeSizes[8] = Rubinius::L64 ? :long : :long_long
+
+  # Load all the platform dependent types
+
+  # MODIFIED
```
Comment (nirvdrum) [inline truffle/src/main/ruby/core/rubinius/platform/ffi.rb]: If we need to diverge from Rubinius, it'd probably be best to document why.

File: truffle/src/main/ruby/core/rubinius/common/process.rb (hunk 2 above)
Comment (eregon) [inline truffle/src/main/ruby/core/rubinius/common/process.rb]: Which JRE do not have Unsafe?

**Comment by eregon** [comment]
I commented per commit as this was easier.
Looks good, I'll merge it!

**Comment by eregon** [comment]
I'll let you merge, in case you want toa dd additional commits before the merge.

**Comment by chrisseaton** [comment]
@nirvdrum we're using jnr-posix to make these calls, and there is a Windows and generic Java implementation of most of them. For example `unlink` simply does `File.delete` on a JVM without JNI or on Windows. I hope that covers us for all the POSIX calls that make sense. I guess things like `lstat` aren't implemented for non-POSIX systems, but then what would it do on Windows anyway.

File: truffle/src/main/ruby/core/rubinius/common/process.rb (hunk 2 above)
Comment (chrisseaton) [inline truffle/src/main/ruby/core/rubinius/common/process.rb]: I think Truffle requires it for the object model anyway, so the decision has been made for us.

File: truffle/src/main/ruby/core/rubinius/platform/ffi.rb (hunk 3 above)
Comment (chrisseaton) [inline truffle/src/main/ruby/core/rubinius/platform/ffi.rb]: Yes sorry I must have been in the middle of a thought here ands stopped mid-comment.

**Pull request merged**